src/tactics.ml

```diff
@@ -103,6 +103,42 @@ let handle_tactic : sig_state -> Proof.t -> p_tactic -> Proof.t =
       handle_refine (Rewrite.reflexivity tac.pos ps)
   | P_tac_sym           ->
       handle_refine (Rewrite.symmetry tac.pos ps)
+  | P_tac_why3(s)    ->
+      (* get the goal to prove *)
+      let (hypothesis, trm) = Proof.Goal.get_type g in
+      (* get the default or the indicated name of the prover. *)
+      let prover_name =
+        match s with
+        | None          -> Timed.(!Why3prover.current_prover)
+        | Some(name)    -> name
+      in
+      (* translate from lambdapi to why3 terms. *)
+      let (constants_table, hyps, why3term) =
+          Why3prop.translate tac.pos ps.proof_builtins (hypothesis, trm) in
+      (* create a new task that contains symbols, axioms and the goal. *)
+      let tsk = Why3task.create constants_table hyps why3term in
+      (* call the prover named [prover_name] and get the result. *)
+      let prover_result = Why3prover.call tac.pos prover_name tsk in
+      (* if the prover succeed to prover the goal. *)
+      if Why3prover.answer prover_result.pr_answer then
+        (* create a new axiom that represents the proved goal. *)
+        let why3_axiom = Pos.make tac.pos (Why3prop.get_newname ()) in
+        (* get the meta type of the current goal (with quantified context) *)
+        let trm = Timed.(!((Proof.Goal.get_meta g).meta_type)) in
+        (* add the axiom to the current signature. *)
+        let a = Sign.add_symbol ss.signature Const why3_axiom trm [] in
+        (* tell the user that the goal is proved (verbose 2) *)
+        Console.out 2 "%s proved the current goal@." prover_name;
+        (* return the variable terms of each item in the context. *)
+        let free_var = fun (_, (x, _)) -> Terms.mkfree x in
+        let context = List.rev_map free_var hypothesis in
```

Comment:
Unfold the definition of free_var as it is short.

Reply:
Rename context into terms, because it is not a context (in the sense of Ctxt.t) but a list of terms.